### PR TITLE
Integrate MKL-DNN leakyrelu

### PR DESCRIPTION
**Pull request by xinyu-intel**
## Description ##
Integrate `LeakyRelu` operator with MKL-DNN activation primitives when the `act_type` is `leakyrelu` and `elu`(gelu will be available when #16073 is ready)

Main Change:
- Register `LeakyRelu` operator with NNVM.
- Share MKL-DNN integration code with `Activation` operator.
- Support mkl-dnn `elu` and `leakyrelu`.
- Support `conv + leakyrelu` fusion and quantization, which appears as a common structure in darknet and yolov3.

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

File: src/operator/nn/mkldnn/mkldnn_act.cc [hunk 1]
```diff
@@ -57,30 +57,56 @@ bool SupportMKLDNNAct(const ActivationParam& param, const NDArray &input) {
...
-      return mkldnn::algorithm::eltwise_logistic;
-    case activation::kTanh:
-      return mkldnn::algorithm::eltwise_tanh;
-    case activation::kSoftReLU:
-      return mkldnn::algorithm::eltwise_soft_relu;
-    default:
-      LOG(FATAL) << "unknown activation type";
-      return mkldnn::algorithm::eltwise_relu;
+mkldnn::algorithm GetMKLDNNActAlgo(const MKLDNNActParam& param) {
+  if (param.is_leakyrelu) {
```
Comment (ZhennanQin) [inline src/operator/nn/mkldnn/mkldnn_act.cc]: Why we need to collect `is_leakyrelu` in  `MKLDNNActParam`? I don't see any special for them.

Comment (xinyu-intel) [reply src/operator/nn/mkldnn/mkldnn_act.cc]: addressed

File: src/operator/leaky_relu-inl.h [hunk 2]
```diff
@@ -332,7 +332,50 @@ class LeakyReLUOp : public Operator {
...
+  const std::vector<TBlob> &in_grad = outputs;
+
+  MSHADOW_REAL_TYPE_SWITCH(dtype, DType, {
+    LeakyReLUOp<xpu, DType> op(param);
+    op.Backward(ctx, out_grad, in_data, out_data, req, in_grad, no_use_but_adapt_origin_api);
+  });
+}
 
 #if DMLC_USE_CXX11
 class LeakyReLUProp : public OperatorProperty {
```
Comment (TaoLv) [inline src/operator/leaky_relu-inl.h]: Can be removed?

Comment (xinyu-intel) [reply src/operator/leaky_relu-inl.h]: addressed

File: src/operator/leaky_relu-inl.h [hunk 3]
```diff
@@ -332,166 +332,51 @@ class LeakyReLUOp : public Operator {
...
+  const std::vector<TBlob> no_use_but_adapt_origin_api;
+  // inputs: out_grad, input_data, input_gamma, output, output_mask
+  size_t expected_in = param.act_type == leakyrelu::kPReLU ? 2 : 1;
+  size_t expected_out = param.act_type == leakyrelu::kRReLU ? 2 : 1;
 
-  std::vector<ResourceRequest> BackwardResource(
-      const mxnet::ShapeVector &in_shape) const override {
-    return {ResourceRequest::kTempSpace};
-  }
+  std::vector<TBlob> out_grad{inputs[0]};
```
Comment (TaoLv) [inline src/operator/leaky_relu-inl.h]: Before this, need check `intputs.size() >= 1 + expected_in + expected_out`.

File: src/operator/leaky_relu-inl.h [hunk 4]
```diff
@@ -332,166 +332,51 @@ class LeakyReLUOp : public Operator {
...
+  std::vector<TBlob> out_data(inputs.begin() + 1 + expected_in,
+                              inputs.begin() + 1 + expected_in + expected_out);
 
-  Operator* CreateOperator(Context ctx) const override {
-    LOG(FATAL) << "Not Implemented.";
-    return NULL;
-  }
+  CHECK_EQ(out_grad.size(), 1U);
+  CHECK_EQ(req.size(), expected_in);
+  CHECK_EQ(in_data.size(), expected_in);
```
Comment (TaoLv) [inline src/operator/leaky_relu-inl.h]: No need to check the size of out_grad and in_data. They're created in this function.
`CHECK_EQ(req.size(), outputs.size());`

File: src/operator/leaky_relu.cc [hunk 5]
```diff
@@ -25,27 +25,123 @@
...
+  if (param_.act_type == leakyrelu::kRReLU) {
+    out_shape->push_back(dshape);
+  }
+  return true;
 }
 
-DMLC_REGISTER_PARAMETER(LeakyReLUParam);
+#if MXNET_USE_MKLDNN == 1
+static void LeakyReLUComputeExCPU(const nnvm::NodeAttrs& attrs,
+                                   const OpContext& ctx,
```
Comment (TaoLv) [inline src/operator/leaky_relu.cc]: Please fix the indentation.

File: src/operator/leaky_relu.cc [hunk 6]
```diff
@@ -25,27 +25,123 @@
...
+    MKLDNN_OPCHECK_INIT(false, outputs.size(), inputs, outputs);
+    MKLDNNLeakyReluForward(attrs, ctx, inputs[0], req[0], outputs[0]);
+    MKLDNN_OPCHECK_RUN(LeakyReLUCompute<cpu>, attrs, ctx, inputs, req, outputs);
+    return;
+  }
+  FallBackCompute(LeakyReLUCompute<cpu>, attrs, ctx, inputs, req, outputs);
+}
+
+void LeakyReLUGradComputeExCPU(const nnvm::NodeAttrs& attrs,
+                                const OpContext& ctx,
```
Comment (TaoLv) [inline src/operator/leaky_relu.cc]: Please fix the indentation.

File: src/operator/leaky_relu.cc [hunk 7]
```diff
@@ -25,27 +25,123 @@
...
+                                const std::vector<NDArray>& inputs,
+                                const std::vector<OpReqType>& req,
+                                const std::vector<NDArray>& outputs) {
+  const LeakyReLUParam& param = nnvm::get<LeakyReLUParam>(attrs.parsed);
+  if (SupportMKLDNNLeakyRelu(param, inputs[0])) {
+    MKLDNN_OPCHECK_INIT(true, outputs.size(), inputs, outputs);
+    MKLDNNLeakyReluBackward(attrs, ctx, inputs.at(0), inputs.at(1), req[0],
```
Comment (TaoLv) [inline src/operator/leaky_relu.cc]: Only two inputs are needed? Is it possible to use vector so we can have a more unified interface?

File: src/operator/leaky_relu.cc [hunk 8]
```diff
@@ -25,27 +25,123 @@
...
+                             outputs[0]);
+    MKLDNN_OPCHECK_RUN(LeakyReLUGradCompute<cpu>, attrs, ctx, inputs, req, outputs);
+    return;
+  }
+  FallBackCompute(LeakyReLUGradCompute<cpu>, attrs, ctx, inputs, req, outputs);
+}
+
+inline static bool LeakyReLUStorageType(const nnvm::NodeAttrs& attrs,
+                                         const int dev_mask,
```
Comment (TaoLv) [inline src/operator/leaky_relu.cc]: Indentation.

File: src/operator/leaky_relu.cc [hunk 9]
```diff
@@ -25,27 +25,123 @@
...
+  const LeakyReLUParam& param = nnvm::get<LeakyReLUParam>(attrs.parsed);
+  size_t expected = param.act_type == leakyrelu::kPReLU ? 2 : 1;
+  CHECK_EQ(in_attrs->size(), expected);
+  return MKLDNNStorageType(attrs, dev_mask, SupportMKLDNNLeakyRelu(param),
+                           dispatch_mode, in_attrs, out_attrs);
+}
 
-MXNET_REGISTER_OP_PROPERTY(LeakyReLU, LeakyReLUProp)
+inline static bool BackwardLeakyReLUStorageType(const nnvm::NodeAttrs& attrs,
+                                          const int dev_mask,
```
Comment (TaoLv) [inline src/operator/leaky_relu.cc]: Indentation.

File: src/operator/nn/mkldnn/mkldnn_act.cc [hunk 10]
```diff
@@ -79,18 +93,30 @@ mkldnn::algorithm GetMKLDNNActAlgo(const ActivationParam& param) {
...
 }
 
+mkldnn::algorithm GetMKLDNNActAlgo(const LeakyReLUParam& param) {
+  switch (param.act_type) {
+    case leakyrelu::kLeakyReLU:
+      return mkldnn::algorithm::eltwise_relu;
+    case leakyrelu::kELU:
+      return mkldnn::algorithm::eltwise_elu;
+    default:
+      LOG(FATAL) << "unknown activation type";
```
Comment (TaoLv) [inline src/operator/nn/mkldnn/mkldnn_act.cc]: More descriptive error message"
```suggestion
      LOG(FATAL) << "unknown activation type for LeakyReLU: " << param.act_type;
```

File: src/operator/nn/mkldnn/mkldnn_ops-inl.h [hunk 11]
```diff
@@ -109,6 +109,12 @@ void MKLDNNActivationForward(const nnvm::NodeAttrs& attrs, const OpContext &ctx,
 void MKLDNNActivationBackward(const nnvm::NodeAttrs& attrs, const OpContext &ctx,
                               const NDArray &out_grad, const NDArray &in_data,
                               const OpReqType &req, const NDArray &in_grad);
+void MKLDNNLeakyReluForward(const nnvm::NodeAttrs& attrs, const OpContext &ctx,
+                             const NDArray &in_data, const OpReqType &req,
```
Comment (TaoLv) [inline src/operator/nn/mkldnn/mkldnn_ops-inl.h]: Indentation.

File: tests/python/unittest/test_gluon.py [hunk 12]
```diff
@@ -1222,7 +1222,9 @@ def elu(x):
         return [elu(x_i) for x_i in x]
 
     for test_point, ref_point in zip(elu_test(point_to_validate), elu(point_to_validate)):
-        assert test_point == ref_point
+        print(test_point)
+        print(ref_point)
```
Comment (TaoLv) [inline tests/python/unittest/test_gluon.py]: Why print?

File: src/operator/leaky_relu.cc (hunk 7 above)
Comment (xinyu-intel) [reply src/operator/leaky_relu.cc]: use inputs vector directly and we need first two of three NDArrays, so add a `CHECK_GE(inputs, 2U)`. Is it make sense?

**Comment by pengzhao-intel** [comment]
@xinyu-intel the MKL-DNN is updated to 0.21 now.

**Comment by xinyu-intel** [comment]
@pengzhao-intel I'll start another PR to enable gelu after this PR merged:)

**Comment by pengzhao-intel** [comment]
+ @ZhennanQin

**Comment by pengzhao-intel** [comment]
@xinyu-intel  is this ready to merge?

**Comment by xinyu-intel** [comment]
yes

**Comment by pengzhao-intel** [comment]
Merging now.

**Pull request merged**